### PR TITLE
Add weighted version of TransformerUnlikelihoodAgent

### PR DESCRIPTION
**Pull request by jakeane**
**Patch description**

Added an agent for unlikelihood dialogue, `TransformerWeightedUnlikelihoodAgent` that allows rewards to be weighted as opposed to binary. This involved creating a new trait, `RewardWeightedUnlikelihoodAgentTrait`, that extends the functionality of `RewardUnlikelihoodAgent` such that the loss for each batch example is multiplied by `batch.rewards`.

This addition will allow parts of the training data to be selectively ignored or emphasized, thus allowing the model to be better tailored.

**Testing steps**

Not entirely sure how agents under `projects/dialogue_unlikelihood` are tested, but I ran these two test commands. 
```
pytest -k TestTorchAgent
pytest -k TestTransformer
```
The first ran successfully, but the second failed the tests for `test_wide_distillation_losses` and `test_narrow_distillation_losses` under `tests/nightly/gpu/anti_scaling/test_anti_scaling.py`. *However*, these tests also failed under the `master` branch, so I do not think it is related to my addition.

**Related Issues**

 - [#2966](https://github.com/facebookresearch/ParlAI/issues/2966)
 - [#3507](https://github.com/facebookresearch/ParlAI/issues/3507)

### CODE REVIEW

File: projects/dialogue_unlikelihood/agents.py [hunk 1]
```diff
@@ -535,6 +535,100 @@ def compute_loss(self, batch, return_output=False):
...
+        ul_scores = scores_view[range_, targets_view]
+        clamp_min = 1e-6 if self.opt['fp16'] else 1e-20
+        ul_loss = torch.mv(
+            (
+                torch.log(torch.clamp(1.0 - ul_scores.exp(), min=clamp_min)).view_as(
+                    ul_notnull
+                )
+                * ul_notnull.float()
+            ).T,
+            batch.rewards,
```
Comment (stephenroller) [inline projects/dialogue_unlikelihood/agents.py:618]: I think you want to negate this here

File: projects/dialogue_unlikelihood/agents.py [hunk 2]
```diff
@@ -535,6 +535,100 @@ def compute_loss(self, batch, return_output=False):
...
+            else:
+                return mle_loss
+
+        # and now we want the unlikelihood loss on the negative examples
+        ul_notnull = notnull & (batch.rewards < 0).unsqueeze(1).expand_as(notnull)
+        ul_target_tokens = ul_notnull.long().sum()
+        range_ = torch.arange(targets_view.size(0)).to(batch.label_vec.device)
+        ul_scores = scores_view[range_, targets_view]
+        clamp_min = 1e-6 if self.opt['fp16'] else 1e-20
+        ul_loss = torch.mv(
```
Comment (stephenroller) [inline projects/dialogue_unlikelihood/agents.py:611]: both here and in MLE, it looks like you're doing a matrix-vector multiplication to cast the rewards across the entire sequence. This likely means that the per-token averages of `nll_loss`, `ppl`, and `ul_loss` are actually per-sequence averages. You probably wish to check this.

**Review by stephenroller — COMMENTED** [review]
Left some comments to help your exploration. Given that we don't have an official usage of this right now, I don't think we'll be accepting this upstream at this moment; in order to combat bloat, I try encourage experimental things to stay in forks or internal copies until they've been proven empirically.

Should you present or find reference to strong empirical results on this formulation, then I'll be more than happy to support this, given the addition tests.

In the mean time, I'm happy to continue to review this code to help you implement this for your own experimentation.

File: projects/dialogue_unlikelihood/agents.py (hunk 1 above)
Comment (jakeane) [reply projects/dialogue_unlikelihood/agents.py:618]: I don't think so, as `batch.rewards` would handle the negation instead of adding a `-` in front of `torch.log()` like in `RewardLikelihoodAgentTrait`

File: projects/dialogue_unlikelihood/agents.py (hunk 2 above)
Comment (jakeane) [reply projects/dialogue_unlikelihood/agents.py:611]: The calculation of `mle_target_tokens` and `ul_target_tokens` are unchanged from `RewardLikelihoodAgentTrait`, and when I tested the values of `mle_loss` and `ul_loss` when the rewards are 1 and -1, the calculated values are the same as those in `RewardLikelihoodAgentTrait`. Similarly, when the rewards are 0.5 and -0.5 for example, the calculated values are exactly half of those calculated in `RewardLikelihoodAgentTrait`. So, I am not exactly sure how this would result in `nll_loss`, `ppl`, and `ul_loss` to be per-sequence averages as opposed to per-token averages.
